nav2_amcl/src/amcl_node.cpp

```diff
@@ -937,6 +940,110 @@ AmclNode::initParameters()
   }
 }
 
+void
+AmclNode::mapReceived(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  RCLCPP_DEBUG(get_logger(), "AmclNode: A new map was received.");
+  if (first_map_only_ && first_map_received_) {
+    return;
+  }
+
+  handleMapMessage(*msg);
+  first_map_received_ = true;
+}
+
+void
+AmclNode::handleMapMessage(const nav_msgs::msg::OccupancyGrid & msg)
+{
+  std::lock_guard<std::recursive_mutex> cfl(configuration_mutex_);
+
+  RCLCPP_INFO(get_logger(), "Received a %d X %d map @ %.3f m/pix\n",
+    msg.info.width,
+    msg.info.height,
+    msg.info.resolution);
+  if (msg.header.frame_id != global_frame_id_) {
+    RCLCPP_WARN(
+      get_logger(), "Frame_id of map received:'%s' doesn't match global_frame_id:'%s'. This could"
+      " cause issues with reading published topics",
+      msg.header.frame_id.c_str(),
+      global_frame_id_.c_str());
+  }
+
+  freeMapDependentMemory();
+  // Clear queued laser objects because they hold pointers to the existing
+  // map, #5202.
+  lasers_.clear();
+  lasers_update_.clear();
+  frame_to_laser_.clear();
+
+  map_ = convertMap(msg);
+
+#if NEW_UNIFORM_SAMPLING
```

Comment:
This is duplicated from the `initMap` function isn't it? Can we factor out the duplication?